cmd/kubeadm/app/util/apiclient/idempotency.go

```diff
@@ -37,28 +38,36 @@ import (
 	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
 )
 
-// ConfigMapMutator is a function that mutates the given ConfigMap and optionally returns an error
-type ConfigMapMutator func(*v1.ConfigMap) error
+// objectMutator is a function that mutates the given runtime object and optionally returns an error
+type objectMutator[T runtime.Object] func(T) error
 
 // apiCallRetryInterval holds a local copy of apiCallRetryInterval for testing purposes
 var apiCallRetryInterval = constants.KubernetesAPICallRetryInterval
 
-// TODO: We should invent a dynamic mechanism for this using the dynamic client instead of hard-coding these functions per-type
+type kubernetesInterface[T kubernetesObject] interface {
+	Create(ctx context.Context, obj T, opts metav1.CreateOptions) (T, error)
+	Get(ctx context.Context, name string, opts metav1.GetOptions) (T, error)
+	Update(ctx context.Context, obj T, opts metav1.UpdateOptions) (T, error)
+}
 
-// CreateOrUpdateConfigMap creates a ConfigMap if the target resource doesn't exist. If the resource exists already, this function will update the resource instead.
-func CreateOrUpdateConfigMap(client clientset.Interface, cm *v1.ConfigMap) error {
+type kubernetesObject interface {
+	runtime.Object
+	metav1.Object
+}
+
+// CreateOrUpdate creates a runtime object if the target resource doesn't exist. If the resource exists already, this function will update the resource instead.
+func CreateOrUpdate[T kubernetesObject](ctx context.Context, client kubernetesInterface[T], obj T) error {
 	var lastError error
-	err := wait.PollUntilContextTimeout(context.Background(),
+	err := wait.PollUntilContextTimeout(ctx,
 		apiCallRetryInterval, kubeadmapi.GetActiveTimeouts().KubernetesAPICall.Duration,
-		true, func(_ context.Context) (bool, error) {
-			ctx := context.Background()
-			if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(ctx, cm, metav1.CreateOptions{}); err != nil {
+		true, func(ctx context.Context) (bool, error) {
+			if _, err := client.Create(ctx, obj, metav1.CreateOptions{}); err != nil {
```

Comment:
i missed this part. PollUntilContextTimeout has caused issues around our code base actually WRT error propagation.

we don't want to use the passtrough context, so this should be left 
```
func(_ context.Context) (bool, error) {
    ctx := context.Background()
```

Reply:
Out of curiosity, do you have pointers to the issues related to error propagation?

I’ve updated the PR to avoid passing the context.

Reply:
i fetched it from git blame on the `ctx := ...` lines.
- https://github.com/kubernetes/kubernetes/pull/123271

```
NOTE: The context passed to PollUntilContextTimeout is not propagated
in the polled function. Instead the poll function creates it's own
context 'ctx := context.Background()', this is to avoid
breaking expectations on the side of the callers, that expect
a certain type of error and not "context timeout" errors.
```

https://github.com/kubernetes/kubernetes/pull/123271#discussion_r1492038591

and think this was discovered as an issue while i worked on the above PR.

Reply:
Ah right, thanks — I didn’t think to look here specifically, I thought it might be related to other uses!